implementations/rust/ockam/ockam_identity/src/secure_channel/handshake/handshake.rs

```diff
@@ -0,0 +1,641 @@
+    fn read_message2_encrypted_key(message: &Vec<u8>) -> Result<&[u8]> {
+        Self::read_middle(message, Self::key_size(), Self::encrypted_key_size())
+    }
+
+    /// Read the message 2 encrypted payload, which is present after the encrypted key
+    fn read_message2_payload(message: &Vec<u8>) -> Result<&[u8]> {
+        Self::read_end(message, Self::key_size() + Self::encrypted_key_size())
+    }
+
+    /// Read the message 3 encrypted key at the beginning of the message
+    fn read_message3_encrypted_key(message: &Vec<u8>) -> Result<&[u8]> {
+        Self::read_start(message, Self::encrypted_key_size())
+    }
+
+    /// Read the message 3 payload which is present after the encrypted key
+    fn read_message3_payload(message: &Vec<u8>) -> Result<&[u8]> {
+        Self::read_end(message, Self::encrypted_key_size())
+    }
+
+    /// Read the first 'length' bytes of the message
+    fn read_start(message: &Vec<u8>, length: usize) -> Result<&[u8]> {
+        if message.len() < length {
+            return Err(XXError::MessageLenMismatch.into());
+        }
+        Ok(&message[0..length])
+    }
+
+    /// Read the bytes of the message after the first 'drop_length' bytes
+    fn read_end(message: &Vec<u8>, drop_length: usize) -> Result<&[u8]> {
+        if message.len() < drop_length {
+            return Err(XXError::MessageLenMismatch.into());
+        }
+        Ok(&message[drop_length..])
+    }
+
+    /// Read 'length' bytes of the message after the first 'drop_length' bytes
+    fn read_middle(message: &Vec<u8>, drop_length: usize, length: usize) -> Result<&[u8]> {
+        if message.len() < drop_length + length {
+            return Err(XXError::MessageLenMismatch.into());
+        }
+        Ok(&message[drop_length..(drop_length + length)])
+    }
+
+    /// Read the bytes of a key at the beginning of a message
+    fn read_key(message: &Vec<u8>) -> Result<&[u8]> {
+        Self::read_start(message, Self::key_size())
+    }
+
+    /// Size of a public key
+    fn key_size() -> usize {
+        CURVE25519_PUBLIC_LENGTH_USIZE
+    }
+
+    /// Size of an encrypted key
+    fn encrypted_key_size() -> usize {
+        Self::key_size() + AES_GCM_TAGSIZE_USIZE
+    }
+}
+
+/// The `HandshakeState` contains all the variables necessary to follow the Noise protocol
+#[derive(Debug, Clone)]
+pub(super) struct HandshakeState {
+    pub(super) s: KeyId,
+    e: KeyId,
+    k: KeyId,
+    re: PublicKey,
+    pub(super) rs: PublicKey,
+    n: u16,
+    h: [u8; SHA256_SIZE_USIZE],
+    ck: KeyId,
+    message1_payload: Vec<u8>,
+    pub(super) status: Status,
+}
+
+impl HandshakeState {
+    /// Create a new HandshakeState with:
+    ///   - a static key
+    ///   - an ephemeral key
+    ///   - an encoded identity payload (identity + signature + credentials)
+    pub(super) fn new(s: KeyId, e: KeyId, message1_payload: Vec<u8>) -> HandshakeState {
+        HandshakeState {
+            s,
+            e,
+            k: "".to_string(),
```

Comment:
Why don't use `Option` for `k`, `re`, `rs`, `ck`? I think `PublicKey` should always contain a valid Public Key. `KeyId` should point to an existent key in the Vault (though this one is hard to guarantee 100% of time)

Reply:
I did this on purpose because I had the intuition that making a stricter state machine where we truly prevent incorrect states would make the code more complex. I have now attempted to go in that direction with more precise state encoding the exact amount of knowledge that a state machine has: `initial -> k, ck -> re -> re, rs` (and variations of that). 

It indeed makes the code more complicated and IMO harder to review for people who follow the informal algorithm on the side where variables can just be uninitialized. So I propose that we leave the code like this for now.

Reply:
IMHO, this stands against the idea of Rust. I agree that the code may need to become more cluttered in this case, but I see it from a different point of view: you were able to avoid all of that clutter because you relied on assumptions about your code flow. I see multiple problems with that approach:
1. For example, you think there is no need to check if `re` is empty because you know that it should be initialized with a valid value before it's used. But that is a very non-obvious assumption, and the compiler is not able to check. Also, that assumption relies on the assumption about the calling side - because `re` is initialized in `decode_message1` but used from a different function - `encode_message2`.
2. You could be wrong about your assumptions (people make mistakes), which can lead to the usage of uninitialized variables, which is potentially dangerous and unpredictable. Also, the code will be modified sooner or later, and who knows if the person modifying it will be able to easily figure out the assumptions you used and not break them while modifying the code.

I think we should avoid that. If `PublicKey` is a type that is meant to hold a valid public key, we should never break that assumption, in this case, the rust will also guarantee that it stays like that. For example, java will always allow you to put `null` as a value, people used to rely on assumptions in their code and were not checking values for null all the time, which led to `NullPointerException` becoming a meme. In Rust we have the luxury of explicitly distinguishing between variables that could be uninitialized and variables that are guaranteed to be initialized, and we rely on the compiler checking that for us. This is one of the reasons why rust is considered a safe language and one of the key reasons we use it.

On the other hand, I thought that switching to a State machine helps to avoid that problem in the first place because you have a set of states you can be in, and you know what the set of initialized variables in each state is (which also requires some clutter, so what need to evaluate what is better). But it looks like it was the case before that PR (I was not very familiar with the state we were in after my code was modifier but before this PR). So, is that the State machine we wanted to achieve?

Reply:
Done, I replaced uninitialized values with `None` and a proper error message if we were to access such a value.